### PR TITLE
Deprecate fix_plus_plus_recursive()

### PR DESCRIPTION
**Pull request by nrwahl2**

### CODE REVIEW

**Comment by nrwahl2** [comment]
I don't want to deal with this right now, but it doesn't make a lot of sense to me that modify and replace behave differently re: ++ and +=. We could either deprecated that behavior in modify or (eventually) implement it correctly in replace. I kind of doubt anyone is using it, but it does currently work with modify.

**Review by kgaillot — APPROVED** [review]
na na na na hey hey goodbye

**Comment by kgaillot** [comment]
> I don't want to deal with this right now, but it doesn't make a lot of sense to me that modify and replace behave differently re: ++ and +=. We could either deprecated that behavior in modify or (eventually) implement it correctly in replace. I kind of doubt anyone is using it, but it does currently work with modify.
> 
> (I know why they differ; I mean I don't think they should.)

Deprecate

**Comment by nrwahl2** [comment]
Sigh
```
git diff
diff --git a/cts/cli/regression.tools.exp b/cts/cli/regression.tools.exp
index 9b5f400e4b..0972dc142d 100644
--- a/cts/cli/regression.tools.exp
+++ b/cts/cli/regression.tools.exp
@@ -7060,7 +7060,6 @@ Transition Summary:
   * Start      httpd:2            ( httpd-bundle-2 )  due to unrunnable httpd-bundle-docker-2 start (blocked)
 
 Executing Cluster Transition:
-  * Cluster action:  clear_failcount for ping on cluster02
   * Pseudo action:   ping-clone_stop_0
   * Pseudo action:   httpd-bundle_start_0
   * Resource action: ping            stop on cluster02
```

about to look into it...

**Comment by nrwahl2** [comment]
> Sigh
> ...
> about to look into it...

I think the new behavior might be correct? This is from running
```
crm_simulate -x $CIB_file -S -i ping_monitor_10000@cluster02=1
```

That shouldn't clear the fail-count.

However, this does expose an issue in `pcmk__inject_failcount()`.
```
void
pcmk__inject_failcount(pcmk__output_t *out, xmlNode *cib_node,
                       const char *resource, const char *task,
                       guint interval_ms, int rc)
{
    ...
        name = pcmk__failcount_name(resource, task, interval_ms);
        inject_transient_attr(out, cib_node, name, "value++");
    ...
```

Previously, this must have been setting the value to 0 since (as discussed in the commit message) `fix_plus_plus_recursive()` doesn't have access to the original value.

Now, it's setting the value to the literal `"value++"`.

Neither of those behaviors is correct for `pcmk__inject_failcount()`.

**Comment by nrwahl2** [comment]
> Previously, this must have been setting the value to 0 since (as discussed in the commit message) `fix_plus_plus_recursive()` doesn't have access to the original value.

Actually I would've expected it to set the new value to 1; it would treat the initial value as 0 and increment that.

So I'm not sure why there was a `clear_failcount` operation scheduled prior to this commit.

**Comment by kgaillot** [comment]
> > Sigh
> > ...
> > about to look into it...
> 
> I think the new behavior might be correct? This is from running
> 
> ```
> crm_simulate -x $CIB_file -S -i ping_monitor_10000@cluster02=1
> ```
> 
> That shouldn't clear the fail-count.

I thought we do clear the failcount when a resource is stopped, but I may be misremembering

> 
> However, this does expose an issue in `pcmk__inject_failcount()`.
> 
> ```
> void
> pcmk__inject_failcount(pcmk__output_t *out, xmlNode *cib_node,
>                        const char *resource, const char *task,
>                        guint interval_ms, int rc)
> {
>     ...
>         name = pcmk__failcount_name(resource, task, interval_ms);
>         inject_transient_attr(out, cib_node, name, "value++");
>     ...
> ```
> 
> Previously, this must have been setting the value to 0 since (as discussed in the commit message) `fix_plus_plus_recursive()` doesn't have access to the original value.
> 
> Now, it's setting the value to the literal `"value++"`.
> 
> Neither of those behaviors is correct for `pcmk__inject_failcount()`.

Yeah let's fix it

**Comment by kgaillot** [comment]
> Actually I would've expected it to set the new value to 1; it would treat the initial value as 0 and increment that.
> 
> So I'm not sure why there was a `clear_failcount` operation scheduled prior to this commit.

That makes sense then, if previously it became 1, then it needed to be cleared for the stop; `char2score()` would now treat the value as 0, so it doesn't need to be cleared. Once we fix the injection it should schedule the cleanup again.

**Comment by nrwahl2** [comment]
> I thought we do clear the failcount when a resource is stopped

That's quite surprising to me, but the other things make sense in light of it, so thank you.

**Comment by nrwahl2** [comment]
That was annoying.

Passes on my machine, here's hoping it passes CI.

**Comment by nrwahl2** [comment]
> > I thought we do clear the failcount when a resource is stopped
> 
> That's quite surprising to me, but the other things make sense in light of it, so thank you.

I'm not seeing where that's happening. I'm looking for `pe__clear_failcount()` and `PCMK_ACTION_CLEAR_FAILCOUNT`. If you know off the top of your head, I am curious, but don't spend significant effort looking for it.

File: lib/pacemaker/pcmk_injections.c [hunk 1]
```diff
@@ -89,30 +90,46 @@ inject_transient_attr(pcmk__output_t *out, xmlNode *cib_node,
...
-        return;
+    int failcount = 0;
+    xmlNode *output = NULL;
 
-    } else {
-        char *name = NULL;
-        char *now = pcmk__ttoa(time(NULL));
+    CRM_CHECK(cib_node != NULL, return);
+
+    if ((rc == 0) || ((rc == 7) && (interval_ms == 0))) {
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_injections.c:105]: These are `PCMK_OCF_{OK,NOT_RUNNING}`

File: lib/pacemaker/pcmk_injections.c [hunk 2]
```diff
@@ -89,30 +90,46 @@ inject_transient_attr(pcmk__output_t *out, xmlNode *cib_node,
...
 
-    } else if ((rc == 7) && (interval_ms == 0)) {
-        return;
+    int failcount = 0;
+    xmlNode *output = NULL;
 
-    } else {
-        char *name = NULL;
-        char *now = pcmk__ttoa(time(NULL));
+    CRM_CHECK(cib_node != NULL, return);
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_injections.c:103]: all of the pointer arguments have to be non-NULL

File: lib/pacemaker/pcmk_injections.c [hunk 3]
```diff
@@ -89,30 +90,46 @@ inject_transient_attr(pcmk__output_t *out, xmlNode *cib_node,
...
+    if (cib__get_node_attrs(out, cib_conn, PCMK_XE_STATUS,
+                            pcmk__xe_id(cib_node), NULL, NULL, NULL, name,
+                            NULL, &output) == pcmk_rc_ok) {
 
-        free(now);
+        if (crm_element_value_int(output, name, &failcount) != 0) {
+            failcount = 0;
+        }
     }
+    value = crm_strdup_printf("%d", failcount + 1);
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_injections.c:120]: or equivalently `pcmk__itoa(failcount + 1)`

**Review by kgaillot — COMMENTED** [review]
looks good

File: lib/pacemaker/pcmk_injections.c (hunk 1 above)
Comment (kgaillot) [reply lib/pacemaker/pcmk_injections.c:105]: and while we're here, s/rc/exit_status/

File: lib/pacemaker/pcmk_injections.c (hunk 2 above)
Comment (nrwahl2) [reply lib/pacemaker/pcmk_injections.c:103]: whoops

File: lib/pacemaker/pcmk_injections.c (hunk 3 above)
Comment (nrwahl2) [reply lib/pacemaker/pcmk_injections.c:120]: Oh that's right

**Comment by nrwahl2** [comment]
updated per review

**Review by kgaillot — APPROVED** [review]

**Pull request merged**